service/matching/liveness.go

```diff
@@ -36,90 +36,76 @@ type (
 		status     int32
 		timeSource clock.TimeSource
 		ttl        time.Duration
-		// internal shutdown channel
-		shutdownChan chan struct{}
+
+		// stopCh is used to signal the liveness to stop
+		stopCh chan struct{}
+		// wg is used to wait for the liveness to stop
+		wg sync.WaitGroup
 
 		// broadcast shutdown functions
 		broadcastShutdownFn func()
 
-		sync.Mutex
-		lastEventTime time.Time
+		lastEventTimeNano int64
 	}
 )
 
 var _ common.Daemon = (*liveness)(nil)
 
-func newLiveness(
-	timeSource clock.TimeSource,
-	ttl time.Duration,
-	broadcastShutdownFn func(),
-) *liveness {
+func newLiveness(timeSource clock.TimeSource, ttl time.Duration, broadcastShutdownFn func()) *liveness {
 	return &liveness{
-		status:       common.DaemonStatusInitialized,
-		timeSource:   timeSource,
-		ttl:          ttl,
-		shutdownChan: make(chan struct{}),
-
+		status:              common.DaemonStatusInitialized,
+		timeSource:          timeSource,
+		ttl:                 ttl,
+		stopCh:              make(chan struct{}),
 		broadcastShutdownFn: broadcastShutdownFn,
-
-		lastEventTime: timeSource.Now().UTC(),
+		lastEventTimeNano:   timeSource.Now().UnixNano(),
 	}
 }
 
 func (l *liveness) Start() {
-	if !atomic.CompareAndSwapInt32(
-		&l.status,
-		common.DaemonStatusInitialized,
-		common.DaemonStatusStarted,
-	) {
+	if !atomic.CompareAndSwapInt32(&l.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
 		return
 	}
 
+	l.wg.Add(1)
 	go l.eventLoop()
 }
 
 func (l *liveness) Stop() {
-	if !atomic.CompareAndSwapInt32(
-		&l.status,
-		common.DaemonStatusStarted,
-		common.DaemonStatusStopped,
-	) {
+	if !atomic.CompareAndSwapInt32(&l.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
 		return
 	}
 
-	close(l.shutdownChan)
+	close(l.stopCh)
 	l.broadcastShutdownFn()
+	l.wg.Wait()
```

Comment:
I guess this will help make this block upstream callers know when this is done?

Reply:
Desirable pattern is to finish all cleanup and termination of goroutines before returning from `Stop()`. Also stop child components. If each component does this we will be in very good shape.